bundles/org.eclipse.swt/Eclipse SWT/gtk/org/eclipse/swt/widgets/Tree.java

```diff
@@ -4319,6 +4319,80 @@ void checkSetDataInProcessBeforeRemoval() {
 	}
 }
 
+boolean initPixbufSize(Image image) {
+	if (pixbufSizeSet || image == null) {
```

Comment:
As I understand from the swt code, for any `Tree`:
1. all images in the tree must have the same size
2. that fixed size is the size of the 1st image set for the tree

In the code this is implemented in `TreeItem.setImage(int index, Image image)` [here](https://github.com/eclipse-platform/eclipse.platform.swt/blob/master/bundles/org.eclipse.swt/Eclipse%20SWT/gtk/org/eclipse/swt/widgets/TreeItem.java#L1508-L15189):
```java
	if (image != null) {
		ImageList imageList = parent.imageList;
		if (imageList == null) imageList = parent.imageList = new ImageList();
		int imageIndex = imageList.indexOf(image);
		// When we create a blank image surface gets created with dimensions 0, 0.
        // This call recreates the surface with correct dimensions
		long tempSurface = ImageList.convertSurface(image);
		Cairo.cairo_surface_destroy(tempSurface);
		if (imageIndex == -1) {
			imageIndex = imageList.add(image);
		}
		surface = imageList.getSurface(imageIndex);
		pixbuf = ImageList.createPixbuf(surface);
	}
```
Basically, `ImageList` has `width` and `height` which it initializes from the 1st added image, and then all subsequent added images are scaled to that size.

Here is a snippet that demonstrates how the images in a tree are scaled to the same size:
```java
import org.eclipse.swt.SWT;
import org.eclipse.swt.graphics.*;
import org.eclipse.swt.layout.*;
import org.eclipse.swt.widgets.*;

public class TryColumnsAndImageSizes {

	private Display display;
	private Shell shell;
	private Composite buttonBox;
	private Tree tree;

	private Image[] images = new Image[3];

	public static void main (String[] args) {
		new TryColumnsAndImageSizes ().run ();
	}

	void run () {
		display = new Display ();
		shell = new Shell (display);
		shell.setLayout (new GridLayout ());

		buttonBox = new Composite (shell, 0);
		buttonBox.setLayout (new FillLayout ());

		addSetImageButton ("r50", 0, 20, display.getSystemColor (SWT.COLOR_RED));
		addSetImageButton ("g70", 1, 70, display.getSystemColor (SWT.COLOR_GREEN));
		addSetImageButton ("b90", 2, 120, display.getSystemColor (SWT.COLOR_BLUE));

		tree = new Tree (shell, SWT.VIRTUAL);
		tree.setLayoutData (new GridData (GridData.FILL_BOTH));
		for (int i = 1; i <= 3; i++) {
			var column = new TreeColumn (tree, SWT.LEFT);
			column.setText ("Column " + i);
			column.setWidth (200);
		}

		tree.addListener (SWT.SetData, e -> {
			TreeItem item = (TreeItem) e.item;
			item.setText (0, "a");
			item.setText (1, "b");
			item.setText (2, "c");
			var idx = tree.indexOf (item);
			for (var i = 0; i < images.length; i++) {
				if (idx == i + 2) {
					item.setImage (i, images[i]);
				}
			}
		});
		tree.setItemCount (10);

		shell.setSize (400, 300);
		shell.open ();

		while (!shell.isDisposed ()) {
			if (!display.readAndDispatch ()) {
				display.sleep ();
			}
		}
		display.dispose ();
	}

	private void addSetImageButton (String label, int imageIdx, int imageSize, Color imageColor) {
		var button = new Button (buttonBox, SWT.PUSH);
		button.setText (label);
		button.addListener (SWT.Selection, e -> {
			images[imageIdx] = createImage (imageSize, imageColor);
			tree.clearAll (false);
		});
	}

	private Image createImage (int imageSize, Color color) {
		var image = new Image (display, imageSize, imageSize);
		var gc = new GC (image);
		gc.setForeground (color);
		gc.setBackground (display.getSystemColor (SWT.COLOR_WHITE));
		gc.setLineWidth (5);
		gc.fillRectangle (0, 0, imageSize, imageSize);
		gc.drawRectangle (0, 0, imageSize - 1, imageSize - 1);
		gc.dispose ();
		return image;
	}
}
```

Reply:
As a result, the answer to this:
> Can't we monotonically increase the size for every new image?

is "no, because `Tree` in swt is designed to have images of the same size".

Reply:
Ok, as long as they are not cropped, we can leave this as is.

Reply:
> is "no, because `Tree` in swt is designed to have images of the same size".

It does not matter, as on each new image we could resize all previously added ones as necessary.

Reply:
OK.  
Apparently, this is another issue.  
Just be careful, if you are going to increase fixed row size multiple times, then you'll probably have to replace `fixed_height_mode + gtk_tree_view_row_changed` to `gtk_tree_view_set_model()`. Here is more [details](https://github.com/eclipse-platform/eclipse.platform.swt/pull/1662#issuecomment-2545220394).